### PR TITLE
SOC-1933 | Registration modal broken on pt wikis

### PR DESCRIPTION
**Pull request by Bloodmoon24**
https://wikia-inc.atlassian.net/browse/SOC-1933
@bkoval @katanka  CR please

### CODE REVIEW

**Comment by rogatty** [comment]
This fixes the issue for pt-br but maybe we should fall back to english when there is no configuration for language?

**Comment by kvas-damian** [comment]
agree with @rogatty

**Comment by bkoval** [comment]
+1 to a graceful fallback :)

**Comment by Bloodmoon24** [comment]
@rogatty @kvas-damian @bkoval Thanks for that hint!

File: server/app/lib/AuthUtils.js [hunk 1]
```diff
@@ -92,3 +93,11 @@ export function getWhoAmIUrl() {
 		pathname: localSettings.whoAmIService.path
 	});
 }
+
+/**
+ * @param {string} lang
+ * @returns {string}
+ */
+export function isLanguageDefined(lang) {
```
Comment (Warkot) [inline server/app/lib/AuthUtils.js]: Name of this function is confusing. `is` suggests it returns an answer to a yes/no question → `boolean`. I would rename it to something like `getLanguageWithDefault`, so it explicitly says what happens inside and what expect to be returned.

Comment (bognix) [inline server/app/lib/AuthUtils.js]: :+1:

File: server/app/lib/AuthUtils.js [hunk 2]
```diff
@@ -92,3 +93,11 @@ export function getWhoAmIUrl() {
 		pathname: localSettings.whoAmIService.path
 	});
 }
+
+/**
+ * @param {string} lang
+ * @returns {string}
+ */
+export function getLanguageWithDefault(lang) {
+	return (authLocaleSettings.hasOwnProperty(lang)) ? lang : 'en';
```
Comment (bkoval) [inline server/app/lib/AuthUtils.js]: No need for the additional parenthesis around hasOwnProperty check

File: server/app/lib/AuthUtils.js [hunk 3]
```diff
@@ -92,3 +93,11 @@ export function getWhoAmIUrl() {
 		pathname: localSettings.whoAmIService.path
 	});
 }
+
+/**
+ * @param {string} lang
+ * @returns {string}
+ */
+export function getLanguageWithDefault(lang) {
```
Comment (bkoval) [inline server/app/lib/AuthUtils.js]: lang = 'en'

File: server/app/facets/auth/BirthdateInput.js [hunk 4]
```diff
@@ -43,10 +44,11 @@ export default class BirthdateInput {
 	 * @returns {void}
 	 */
 	constructor(endian, lang) {
-		const langSettings = authLocaleSettings[lang];
-
 		this.endian = endian;
-		this.lang = lang;
+		this.lang = getLanguageWithDefault(lang);
+
+		const langSettings = authLocaleSettings[this.lang];
```
Comment (bkoval) [inline server/app/facets/auth/BirthdateInput.js]: Use const definition at the beginning of the method :)

File: server/app/lib/AuthUtils.js (hunk 3 above)
Comment (Bloodmoon24) [inline server/app/lib/AuthUtils.js]: Nope, It will not work here. We have to check is there translation which you demand, if no, serve en

**Comment by Bloodmoon24** [comment]
@tomasznapieralski Thanks for fixing linter!

File: server/app/facets/auth/BirthdateInput.js [hunk 5]
```diff
@@ -43,10 +44,11 @@ export default class BirthdateInput {
 	 * @returns {void}
 	 */
 	constructor(endian, lang) {
-		const langSettings = authLocaleSettings[lang];
+		let langSettings;
 
 		this.endian = endian;
-		this.lang = lang;
+		this.lang = getLanguageWithDefault(lang);
```
Comment (bkoval) [inline server/app/facets/auth/BirthdateInput.js]: Does this class actually use this property? Or maybe it's just used to get language settings? I cannot find any references to this field in this class...

File: server/app/facets/auth/register.js [hunk 6]
```diff
@@ -43,7 +43,7 @@ import deepExtend from 'deep-extend';
  * @returns {DefaultRegistrationContext}
  */
 function getDefaultRegistrationContext(request, i18n) {
-	const lang = i18n.lng();
+	const lang = authUtils.getLanguageWithDefault(i18n.lng());
```
Comment (bkoval) [inline server/app/facets/auth/register.js]: Seems like obtaining lng() is a repeatable action - maybe we could pass the whole i18n object to the method and call the lng() method inside it?

File: server/tests/unit/lib/AuthUtils.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
+/* globals getLanguageWithDefault:true */
+
+QUnit.module('lib/AuthUtils');
+
+QUnit.test('getLanguageWithDefault', function (assert) {
+	var testCases = [
```
Comment (bkoval) [inline server/tests/unit/lib/AuthUtils.js]: You could add more edge cases here, like passing a variable of a wrong type (any non-string, undefined, null etc.)

File: server/app/lib/AuthUtils.js (hunk 3 above)
Comment (bkoval) [inline server/app/lib/AuthUtils.js]: We could add it anyway, to fall back to english when there's no language provided.

File: server/tests/unit/lib/AuthUtils.js (hunk 7 above)
Comment (Warkot) [inline server/tests/unit/lib/AuthUtils.js]: IMO, you should mock `authLocaleSettings`, to check just if the `getLanguageWithDefault` functions correctly and not if a language (in the case below `ja`) is defined in the `authLocaleSettings`.

You could use `beforeEach` and `afterEach` hooks like [here](https://github.com/Wikia/mercury/blob/dev/front/main/tests/unit/routes/article-test.js#L7-L17) and then set `authLocaleSettings` in every test case like I propose below.

```
var testCases = [
		{
			lang: 'ja',
			expected: 'ja'
			authLocaleSettingsMock: {ja: {}}
		},
		{
			lang: 'Quenya',
			expected: 'en'
			authLocaleSettingsMock: {}
		}
	];

	testCases.forEach(function (testCase) {
		authLocaleSettings = testCase.authLocaleSettingsMock;
		assert.equal(getLanguageWithDefault(testCase.lang), testCase.expected);
	});
```

File: server/app/facets/auth/BirthdateInput.js [hunk 8]
```diff
@@ -43,10 +44,11 @@ export default class BirthdateInput {
 	 * @returns {void}
 	 */
 	constructor(endian, lang) {
```
Comment (tomasznapieralski) [inline server/app/facets/auth/BirthdateInput.js]: Have you considered this:
```
constructor(endian, lang) {
		this.endian = endian;
		this.lang = getLanguageWithDefault(lang);

		this.setSettings();
}

setSettings() {
		const langSettings = authLocaleSettings[this.lang];

		this.settings = langSettings ? langSettings.date : authLocaleSettings.en.date;
}
```
and using `dateSetting` instead of just `settings`?

File: server/app/lib/AuthUtils.js [hunk 9]
```diff
@@ -92,3 +93,13 @@ export function getWhoAmIUrl() {
...
 }
+
+/**
+ * @param {object} i18n
+ * @returns {string}
+ */
+export function getLanguageWithDefault(i18n = {}) {
+	const lang = i18n.lng() || 'en';
+
+	return authLocaleSettings.hasOwnProperty(lang) ? lang : 'en';
```
Comment (rafalkalinski) [inline server/app/lib/AuthUtils.js]: is it possible, that authLocaleSettings.hasOwnProperty('en') would be empty?

**Comment by rafalkalinski** [comment]
One comment, except that: LGTM

**Comment by bkoval** [comment]
LGTM
